lib/stdlib/src/rand.erl

```diff
@@ -1315,6 +1316,85 @@ normal_s(Mean, Variance, State0) when 0 =< Variance ->
     {X, State} = normal_s(State0),
     {Mean + (math:sqrt(Variance) * X), State}.
 
+
+-spec shuffle(list()) -> list().
+shuffle(List) ->
+    {ShuffledList, State} = shuffle_s(List, seed_get()),
+    _ = seed_put(State),
+    ShuffledList.
+
+-spec shuffle_s(list(), state()) -> {list(), state()}.
+shuffle_s(List, {#{bits:=_, next:=Next} = AlgHandler, R0} = State)
+  when is_list(List) ->
+    case List of
+        [] ->
+            {List, State};
+        [_] ->
+            {List, State};
```

Comment:
These clauses are shared for both variants, perhaps we should hoist them into function clauses?

Reply:
I want to avoid duplicating the argument check; there are two branches - either `AlgHandler` has got `bits` and `next`, or it has `mask` and `next`.  After that check that optimization for 0 and 1 element lists is done, and before dismantling `State` into parts since with 0 and 1 element lists the `State` need not change.

So I chose to duplicate the `[]` and `[_]` list tests instead of the `AlgHandler` checks in function clauses.